src/Features/Core/Portable/InvertConditional/AbstractInvertConditionalCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,83 @@
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.InvertConditional
+{
+    internal abstract class AbstractInvertConditionalCodeRefactoringProvider<TConditionalExpressionSyntax> 
+        : CodeRefactoringProvider
+        where TConditionalExpressionSyntax : SyntaxNode
+    {
+        protected abstract bool ShouldOffer(TConditionalExpressionSyntax conditional, int position);
+
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var document = context.Document;
+            var span = context.Span;
+            var cancellationToken = context.CancellationToken;
+
+            if (span.Length > 0)
+            {
+                return;
+            }
+
+            var position = span.Start;
+            var conditional = await FindConditionalAsync(document, position, cancellationToken).ConfigureAwait(false);
+
+            if (conditional == null ||
+                position < conditional.Span.Start ||
+                !ShouldOffer(conditional, position))
+            {
+                return;
+            }
+
+            context.RegisterRefactoring(new MyCodeAction(
+                FeaturesResources.Invert_conditional,
+                c => InvertConditionalAsync(document, position, c)));
+        }
+
+        private static async Task<TConditionalExpressionSyntax> FindConditionalAsync(Document document, int position, CancellationToken cancellationToken)
+        {
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var token = root.FindToken(position);
+
+            return token.Parent.FirstAncestorOrSelf<TConditionalExpressionSyntax>();
+        }
+
+        private async Task<Document> InvertConditionalAsync(
+            Document document, int position, CancellationToken cancellationToken)
+        {
+            var conditional = await FindConditionalAsync(document, position, cancellationToken).ConfigureAwait(false);
+
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+            var editor = new SyntaxEditor(root, document.Project.Solution.Workspace);
+
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+            syntaxFacts.GetPartsOfConditionalExpression(conditional,
+                out var condition, out var whenTrue, out var whenFalse);
+
+            editor.ReplaceNode(condition, editor.Generator.Negate(condition, semanticModel, cancellationToken));
+            editor.ReplaceNode(whenTrue, whenFalse.WithTriviaFrom(whenTrue));
```

Comment:
>WithTriviaFrom [](start = 51, length = 14)

I'm not sure we should swap trivias.
```C#
b
? DoA() // do A because ...
: DoB() // do B because ...
```

Reply:
Yes.  Dilemma.  Especially because it's so important for leading/trailing whitespace/newlines.  If you don't do this, everything goes bad on multi-line stuff.  I think i'm ok as is with the assumption that we can't ever know what trivia is for :)

Reply:
Is there a clever solution where formatting trivia could be swapped, but comment trivia could remain?

---
In reply to: [218178366](https://github.com/dotnet/roslyn/pull/29928#discussion_r218178366) [](ancestors = 218178366)

Reply:
> Is there a clever solution where formatting trivia could be swapped, but comment trivia could remain?

i mean... we have total control over trivia.  But it usually isn't pretty :)  it would probably (for trailing trivia) be matching  a pattern ```space* comment* endofline```, and attempting to swap the above/below, keepign the space+endofline, while swapping the comment portions if there.  

I'm personally not feeling it's worth the complexity :)  maybe if someone wanted to contribute it if they didn't like the default behavior...  what do you think?